Feature/datamap tsp

yifan-zhou922 opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

catalinaperalta committed:
initial commit

catalinaperalta committed:
update attr

catalinaperalta committed:
add security scheme

yifan-zhou922 committed:
Update models.tsp

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update models.tsp

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update examples

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

raych1 commented:
@MaryGao is this a know JS script issue?
```
03:54:45.622 cmdout 	[automation_generate.sh] Generate some other files for @azure-rest/azure-analytics-purview-datamap-rest in /mnt/vss/_work/1/s/azure-sdk-for-js/sdk/azure.analytics.purview.datamap/azure-analytics-purview-datamap-rest...
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] Error:
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] An error occurred while run build for typespec project: "specification/purview/Azure.Analytics.Purview.DataMap":
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] Err: Wrong package path /mnt/vss/_work/1/s/azure-sdk-for-js/sdk/azure.analytics.purview.datamap/azure-analytics-purview-datamap-rest;
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] Stderr: "undefined"
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] Stdout: "undefined"
03:54:45.634 cmderr 	[automation_generate.sh] [ERROR] ErrorStack: "undefined"
03:54:45.635 cmderr 	[automation_generate.sh] [ERROR] Cannot read properties of undefined (reading 'module')
03:54:45.641 error C 	Script return with result [failed] code [1] signal [null] cwd [azure-sdk-for-js]: sh .scripts/automation_generate.sh
```

@chunyu3 , can you have a look at dotnet sdk generation failure? Can you comment if it's a spec issue?

MaryGao commented on specification/purview/Azure.Analytics.Purview.DataMap/tspconfig.yaml:
```diff
@@ -0,0 +1,36 @@
+parameters:  
+  "service-dir":
+    default: "sdk/azure.analytics.purview.datamap"
```
Are we going to use the `azure.analytics.purview.datamap` as folder name? this may not apply JS naming convention in SDK repo and also failed our automation validation. Usually it would be `azureanalyticspurviewdatamap`.

weidongxu-microsoft replied:
should be `sdk/purview`

see https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md#sample-project

yifan-zhou922 replied:
Thanks. Updated

MaryGao commented on specification/purview/Azure.Analytics.Purview.DataMap/tspconfig.yaml:
```diff
@@ -0,0 +1,36 @@
+    generateMetadata: true
+    generateTest: true
+    packageDetails:
+      name: "@azure-rest/azure-analytics-purview-datamap-rest"
```
Usually the package name would be `@azure-rest/{service-name}-{module}` e.g: @azure-rest/analytics-purview-datamap

yifan-zhou922 replied:
Thanks. Updated

MaryGao commented on specification/purview/Azure.Analytics.Purview.DataMap/tspconfig.yaml:
```diff
@@ -0,0 +1,36 @@
+    clear-output-folder: true
+    model-namespace: false
+  "@azure-tools/typespec-ts":
+    package-dir: "azure-analytics-purview-datamap-rest"
```
Usually we won't duplicate the `azure-` prefix in sub folder.
```suggestion
    package-dir: "analytics-purview-datamap-rest"
```

yifan-zhou922 replied:
Thanks. Updated. Is it also apply to other languages as well?

MaryGao replied:
I don't think so, Java would require `azure-` AFAIK.

weidongxu-microsoft replied:
@yifan-zhou922
Basically package-dir is the thing that you would review in SDK review meeting. So you should be more clear about this than us...

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 committed:
Update tspconfig.yaml

yifan-zhou922 committed:
update interface name

MaryGao commented:
/azp run

MaryGao commented:
/azp run

yifan-zhou922 committed:
update query parameter to non-requried and correct int mdoel

yifan-zhou922 committed:
correct response status code from 204 to 200

yifan-zhou922 committed:
Create openapi.json

yifan-zhou922 committed:
update typo

yifan-zhou922 committed:
Update openapi.json

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update examples

yifan-zhou922 committed:
update examples

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update examples

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update example

yifan-zhou922 committed:
update businessMetadata

yifan-zhou922 committed:
Update purviewdatamap.json

yifan-zhou922 committed:
Update custom-words.txt

yifan-zhou922 committed:
Update purviewdatamap.json

yifan-zhou922 committed:
update searchoperator filter

yifan-zhou922 committed:
update import examples

yifan-zhou922 committed:
update

yaleiZhao committed:
Update Search Filters (#26837)

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Merge branch 'feature/datamap-tsp' of ssh://github.com/Azure/azure-rest-api-specs into feature/datamap-tsp

yifan-zhou922 committed:
Delete openapi.json

yifan-zhou922 commented:
Hi  @MaryGao , @raych1 , please help check why the sdk generation pipeline fails. 
For the breaking changes, it was a model error we discovered when migrating from openAPI to typeSpec. (No SDK has been released for this version) 
Please help review the PR. Thanks.

raych1 commented:
> Hi @MaryGao , @raych1 , please help check why the sdk generation pipeline fails. For the breaking changes, it was a model error we discovered when migrating from openAPI to typeSpec. (No SDK has been released for this version) Please help review the PR. Thanks.

@MaryGao JS package was generated but failed to build.

kazrael2119 commented:
/azp run

yifan-zhou922 committed:
add @client

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 committed:
Update tspconfig.yaml

yifan-zhou922 committed:
Update main.tsp

yifan-zhou922 commented:
/azp run

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+#suppress "@azure-tools/typespec-azure-core/casing-style" "It follows the rule"
+#suppress "@azure-tools/typespec-azure-core/bad-record-type" "Should use unknown to model Object"
+@doc("Paginated-list, for returning search results.")
+model PList {
```
Is there a better name for this model?

yifan-zhou922 replied:
It is one of the Atlas models.

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  createdBy?: string;
+
+  @doc("The update time of the record.")
+  updateTime?: int64;
```
Same question about using one of the built in typespec time types?

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  orderby?: unknown[];
+
+  @doc("The filter for the search. See examples for the usage of supported filters.")
+  filter?: unknown;
```
Would it make sense to define the filters through different models?

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+
+  #suppress "@azure-tools/typespec-azure-core/no-unknown" "Should use unknown to model Object"
+  @doc("Any object")
+  sort?: unknown;
```
This type and the doc above it are unclear as to what to do with this property

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+
+@discriminator("kind")
+@doc("Base model for search filter")
+model SearchFilter {
```
It's a bit awkward that this model has no properties. @johanste is this an acceptable design for a base model for discriminated types?

yifan-zhou922 replied:
Comment a little bit. We want to define a base model as SearchFilter which can be a filter for typeName, guid, andFilter, orFilter etc. As there are no shared properties for search filter, this model does not contain any properties.

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+using Azure.Core.Traits.Private;
+using Azure.ClientGenerator.Core;
+
+@client({
```
This should be defined in a client.tsp file

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+
+
+@doc("The ApiVersion query parameter.")
+model ApiVersionParameter {
```
This should be unnecessary if you use the Azure.Core.Foundations.Operation

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+}
+
+@doc("The ApiVersion query parameter for DataMap service.")
+model DataMapApiVersionParameter {
```
This looks like a repeat of the above model and again should be removed in favor of using the Azure.Core.Foundations.Operation template

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+}
+
+@doc("Operation for Atlas service")
+op AtlasBaseOperation<
```
instead of defining this template through basic TypeSpec you should leverage the Azure.Core.Foundations.Operation template

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+
+
+@doc("Operation for DataMap service")
+op DataMapBaseOperation<
```
Same comment as above

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+Traits  & NoConditionalRequests & NoRepeatableRequests & NoClientRequestId           
+>;
+
+//alias DataMapResourceOperations = Azure.Core.ResourceOperations<NoClientRequestId & NoConditionalRequests & NoRepeatableRequests, ErrorResponse>;
```
Let's remove this if it is unused

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2084 @@
+  @doc("Get entity header given its GUID.")
+  @route("/atlas/v2/entity/guid/{guid}/header")
+  @get
+  getHeader is AtlasBaseOperation<
```
Can this use a ResourceRead template?

yifan-zhou922 replied:
The api path does not fit into the template.

yifan-zhou922 committed:
Update specification/purview/Azure.Analytics.Purview.DataMap/main.tsp

Co-authored-by: catalinaperalta <catalinaperaltah@hotmail.com>

timotheeguerin commented on specification/purview/Azure.Analytics.Purview.DataMap/main.tsp:
```diff
@@ -0,0 +1,45 @@
+  "2023-09-01",
+}
+
+#suppress "@azure-tools/typespec-azure-core/casing-style" "It follows the rule"
```
The rule is correct the guidelines says to not use acronym upper cased

timotheeguerin commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  @doc("relationship")
+  RELATIONSHIP,
+  @doc("term template")
+  TERM_TEMPLATE,
```
Yes, it accept anything really if you wrap in backtick if it's not a valid identifier

MaryGao commented on specification/purview/Azure.Analytics.Purview.DataMap/tspconfig.yaml:
```diff
@@ -0,0 +1,36 @@
+    generateTest: true
+    packageDetails:
+      name: "@azure-rest/analytics-purview-datamap"
+      description: "Azure.Analytics.Purview.DataMap Service"
```
this should be used in package.json file as description section.

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  updatedBy?: string;
+
+  @doc("The version of the entity.")
+  version?: int64;
```
No particular format.

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  customAttributes?: Record<string>;
+
+  @doc("The GUID of the entity.")
+  guid?: string;
```
In create scenarios, users can use an unassigned guid for example "-1" or "-2". This indicates that this object is newly created and this unassigned guid is useful for creating new entities with relationships.

For other get scenarios, this guid can use uuid type provided by Azure.Core. I will update accordingly.

yifan-zhou922 committed:
updated

yifan-zhou922 committed:
Update models.tsp

MaryGao commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2116 @@
+    },
+    EntityMutationResponse,
+    {},
+    ErrorResponse
```
I would call out that the ErrorResponse shape is not compatible with error response shape defined in azure core  which will be referred in your RpcOperation one in getSampleBusinessMetadataTemplate. 

https://github.com/Azure/typespec-azure/blob/41bf47f801e64fbc35e03b8d1eb8022959834fa9/packages/typespec-azure-core/lib/foundations.tsp#L129

yifan-zhou922 committed:
update error response

yifan-zhou922 committed:
Update models.tsp

yifan-zhou922 committed:
update descriptions

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2160 @@
+
+@doc("An error response from the service")
+@error
+model AtlasErrorResponse {
```
Thanks. Comment more for this scenario. As most of our APIs are Atlas compatible, these error responses are defined by atlas service.

lmazuel replied:
@MaryGao There is no choice here, as the service API has not been designed by Microsoft as @yifan-zhou922 states

yifan-zhou922 committed:
run tsp format

yifan-zhou922 committed:
Create openapi.json

yifan-zhou922 committed:
update

yifan-zhou922 committed:
add api version

yifan-zhou922 committed:
updated

yifan-zhou922 committed:
guid is not required

yifan-zhou922 committed:
update examples

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_Type.json:
```diff
@@ -5,9 +5,11 @@
           {
-            "entityType": "azure_blob_path",
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path",
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_Type.json:
```diff
@@ -5,9 +5,11 @@
         "and": [
           {
-            "entityType": "azure_blob_path",
+            "kind": "TypeFilter",
```
EntityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_SystemTime.json:
```diff
@@ -5,14 +5,17 @@
+        "kind": "OrFilter",
         "or": [
           {
+            "kind": "CreateTimeFilter",
```
Union

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -2158,11 +2158,25 @@ model ScanInfoFilter extends SearchFilter {
   glossaryType: string;
 }
 
+@doc("Glossary term filter for search")
```
attributeValue type is object

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_And.json:
```diff
@@ -5,11 +5,14 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_AndOrNested.json:
```diff
@@ -5,23 +5,29 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_Not.json:
```diff
@@ -5,8 +5,10 @@
         "not": {
-          "classification": "MICROSOFT.SYSTEM.TEMP_FILE"
+          "kind": "ClassificationFilter",
+          "classificationName": "MICROSOFT.SYSTEM.TEMP_FILE"
```
classification

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_PaginationContinuationPage.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_PaginationLastPage.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Query_Taxonomy.json:
```diff
@@ -25,8 +31,10 @@
             "not": {
-              "classification": "MICROSOFT.SYSTEM.TEMP_FILE"
+              "kind": "ClassificationFilter",
+              "classificationName": "MICROSOFT.SYSTEM.TEMP_FILE"
```
classification

yaleiZhao commented on specification/purview/Azure.Analytics.Purview.DataMap/examples/2023-09-01/Discovery_Suggest.json:
```diff
@@ -4,9 +4,11 @@
           {
-            "entityType": "azure_blob_path",
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path",
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_AutoComplete.json:
```diff
@@ -4,9 +4,11 @@
           {
-            "entityType": "azure_blob_path",
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path",
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_AndOrNested.json:
```diff
@@ -5,23 +5,29 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_Not.json:
```diff
@@ -5,8 +5,10 @@
         "not": {
-          "classification": "MICROSOFT.SYSTEM.TEMP_FILE"
+          "kind": "ClassificationFilter",
+          "classificationName": "MICROSOFT.SYSTEM.TEMP_FILE"
```
classification

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_PaginationContinuationPage.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_PaginationFirstPage.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_PaginationLastPage.json:
```diff
@@ -4,12 +4,15 @@
           {
-            "entityType": "azure_blob_path"
+            "kind": "TypeFilter",
+            "typeName": "azure_blob_path"
```
entityType

yaleiZhao commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/examples/Discovery_Query_Taxonomy.json:
```diff
@@ -25,8 +31,10 @@
             "not": {
-              "classification": "MICROSOFT.SYSTEM.TEMP_FILE"
+              "kind": "ClassificationFilter",
+              "classificationName": "MICROSOFT.SYSTEM.TEMP_FILE"
```
classification

yifan-zhou922 committed:
update search filter

yifan-zhou922 committed:
update model name

yifan-zhou922 committed:
Merge branch 'main' into feature/datamap-tsp

yifan-zhou922 committed:
updated

msyyc commented:
/azp run

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2254 @@
+  contentType: "multipart/form-data";
+}>;
+
+@operationGroup
```
This also belongs in client.tsp

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
Take a look through these docs to create the client.tsp file and to help you define the clients correctly for the SDKs: https://azure.github.io/typespec-azure/docs/howtos/DataPlane%20Generation%20-%20DPG/02client

yifan-zhou922 committed:
update client.tsp

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
update client.tsp

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
update client.tsp

yifan-zhou922 commented:
/azp run

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/main.tsp:
```diff
@@ -0,0 +1,45 @@
+
+#suppress "@azure-tools/typespec-azure-core/casing-style" "It follows the rule"
+@doc("The Azure Active Directory OAuth2 Flow")
+model AADToken
```
To remove the need for the suppression you could update this to:
```suggestion
model AadToken
```

Or call it `AuthToken` or `AzureAuth`

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+""")
+  @route("/atlas/v2/types/typedefs")
+  @post
+  bulkCreate is Azure.Core.Foundations.Operation<
```
```
  bulkCreateTypeDefinitions is Azure.Core.Foundations.Operation<
```

catalinaperalta replied:
@johanste was this a suggestion for the SDK?

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update main.tsp

msyyc commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+  importBusinessMetadata is AtlasImportOperation<
+    {
+      @doc("InputStream of file")
+      uploadedInputStream: bytes;
```
`multipart/form-data` is usually for scenario where payload is mixed with bytes and string. So if you only want to upload bytes, there is no need to use multipart. Of course, if this service plan to add other part type in payload, it is ok to define it in advance.

yifan-zhou922 replied:
Thanks for the information. This is one of the Atlas API https://atlas.apache.org/api/v2/resource_EntityREST.html#resource_EntityREST_importBMAttributes_POST. So we keep it as multipart/form-data.

weidongxu-microsoft replied:
At present, Java won't handle the anonymous model on multipart. Next release would just have it a model `importBusinessMetadataRequest`.

MaryGao commented:
/azp run

yifan-zhou922 commented:
Issues: https://github.com/Azure/azure-sdk-pr/issues/1668

msyyc commented on specification/purview/Azure.Analytics.Purview.DataMap/client.tsp:
```diff
@@ -0,0 +1,153 @@
+  name: "EntityClient",
+  service: PurviewDataMap,
+})
+interface EntityClient {
```
For python, operation group is better than sub client. Here is old package API: https://github.com/Azure/azure-sdk-for-python/blob/cb97ec0a23ebd59eb8fd6d5c338efc25026490ee/sdk/purview/azure-purview-catalog/azure/purview/catalog/_client.py#L71-L91

yifan-zhou922 committed:
Rename models by comments

rename xxContent to xxOptions
rename ContactBasic
rename SearchRequsetTaxonomySetting

yifan-zhou922 committed:
Update openapi.json

yifan-zhou922 committed:
client/method renaming

list->get
bulk -> batch
remove unique
using operation group
guid -> id

msyyc commented on specification/purview/Azure.Analytics.Purview.DataMap/client.tsp:
```diff
@@ -0,0 +1,138 @@
+@TypeSpec.Versioning.useDependency(PurviewDataMap.ApiVersions.v2023_09_01)
+namespace Customizations;
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+@client({
+  name: "DataMapClient",
+  service: PurviewDataMap,
+})
+@operationGroup
```
According to https://azure.github.io/typespec-azure/docs/howtos/DataPlane%20Generation%20-%20DPG/02client#one-client-and-two-operation-groups, `@client` part shall be put for namespace.

msyyc commented on specification/purview/Azure.Analytics.Purview.DataMap/client.tsp:
```diff
@@ -0,0 +1,138 @@
+}
+
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+interface Relationship {
```
Lack `@operationGroup`

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
fix casing problem

yifan-zhou922 committed:
update comments

yifan-zhou922 committed:
update

msyyc commented:
/azp run

yifan-zhou922 committed:
update discovery

yifan-zhou922 committed:
update

yifan-zhou922 committed:
fix example bug

yifan-zhou922 committed:
update partalUpdates

yifan-zhou922 committed:
Merge branch 'main' into feature/datamap-tsp

yifan-zhou922 committed:
update query parameter overwrite

yifan-zhou922 committed:
Update attr to attribute

yifan-zhou922 committed:
Update partialUpdates

yifan-zhou922 committed:
align using body parameter

yifan-zhou922 committed:
Update Glossary_PartialUpdateCategory.json

yifan-zhou922 committed:
revert update examples in original example folder

yifan-zhou922 committed:
Revert "revert update examples in original example folder"

This reverts commit 867767b5dc69f259596e6379aa33617dc6b56d7f.

yifan-zhou922 committed:
Update Glossary_PartialUpdateCategory.json

yifan-zhou922 committed:
Revert "Update Glossary_PartialUpdateCategory.json"

This reverts commit d31fe8aa07e5076cf6e3299c3a9123681ebabd23.

yifan-zhou922 committed:
Revert "Revert "revert update examples in original example folder""

This reverts commit 8f7003247f3d435dc8c80188cb8a5ade814aa137.

yifan-zhou922 committed:
Delete openapi.json

yifan-zhou922 committed:
fix typeSpec validation error

yifan-zhou922 committed:
Update purviewdatamap.json

yifan-zhou922 committed:
Update purviewdatamap.json

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

mikekistler reviewed (changes requested):
There are two concerns with this PR:
1. Azure's versioning policy requires that any change to a GA API version be made in a new api-version. Things like changing parameter names (except path parameters) or making required parameters optional must be done in a new API version.
2. Some of the changes being made appear to be breaking changes from a prior GA, and these must be avoided if at all possible.

yifan-zhou922 commented:
> There are two concerns with this PR:
> 
> 1. Azure's versioning policy requires that any change to a GA API version be made in a new api-version. Things like changing parameter names (except path parameters) or making required parameters optional must be done in a new API version.
> 2. Some of the changes being made appear to be breaking changes from a prior GA, and these must be avoided if at all possible.

Hi @mikekistler, thanks for the response.
All the changes are related to renaming in SDK models. No changes have been done in service API behavior. The purpose for this PR is to migrate to TypeSpec to release SDKs for the existing GA API version (No SDK released yet for 2023-09-01). According to the SDK review comments, we updated some parameter names in TypeSpec and changed the samples accordingly. However, as the previous purviewdatamap.json was using the same set of examples to do the model validation, we have to update the existing swagger file to pass the model validation error, thus causing the breaking changes. 

We did not remove/add any parameters, nor did we change the optional setting for any parameters. All the changes are related to renaming. Below are the justifications we have for those breaking changes. Please help review the breaking changes justifications and provide some suggestions. Many thanks.

- We rename the path parameter name from xxxGuid to xxxId for categoryGuid/glossaryGuid/operationGuid/termGuid [This cause the errors for 1007-RemoveClientParameter and 1038-AddedPath]

    - for example: /atlas/v2/glossary/term/{termId}
- We rename the request body parameter to "body" instead of different model names. [This cause the errors for [1009 - RemovedRequiredParameter](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1009.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801431943%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=718PbS%2FFnlI83FJV9sqofU3bNHi%2F59MDg%2FehAjf5rNw%3D&reserved=0) / [1010 - AddingRequiredParameter](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1010.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801441521%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=ilvXHGTc6Xgpgn9YIVEOFwmFwSAsIsxOPCGrxj%2BuXbo%3D&reserved=0)/ [1043 - AddingOptionalParameter](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1043.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801448927%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=2d3clA3r4qgbQCbqnUy%2BQ%2B2R6zX2LV98nCC6e%2FBl%2FG8%3D&reserved=0) /[1046 - RemovedOptionalParameter](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1046.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801455997%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=ZK15jVveSAnAmchKxrMw3udN89QGh7ThBXKk%2Bve%2BnoY%3D&reserved=0)]
- We use enum "ENUM,ENTITY,CLASSIFICATION,RELATIONSHIP,STRUCT/etc" instead of "enum, entity, classification/etc" The service side is case-insensitive. We update to capital letters to align with the model TypeCategory. [This cause the errors for [1019 - RemovedEnumValue](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1019.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801462745%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=SLbLtFdOWgaewFP%2BLA18WyxUOF1ii2sLL%2BjnBFO%2BpIU%3D&reserved=0)/[1020 - AddedEnumValue](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1020.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801469661%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=EDQeqFu8aGCyJD8KZUbGcLMpFbHaFzTtUi9jAdFSzWM%3D&reserved=0) ]
- For [1038 - AddedPath](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fopenapi-diff%2Ftree%2Fmaster%2Fdocs%2Frules%2F1038.md&data=05%7C02%7Cyifanzhou%40microsoft.com%7C210accacaac94abe893d08dc136cea21%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638406605801476489%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=wEJYjCtIlNRHfcWoBEvm8tcWTG2txb5kkgd8Z7X56MU%3D&reserved=0), there are some false alarming. These paths exist in both purviewdatamap.json and openapi.json

- For 1026-TypeChanged, this is a bug we found in previous swagger when we try to migrate to typeSpec


Here is the feedback we received from previous SDK review meeting. [Azure SDK Review - [Introduction to Microsoft Purview] · Issue #1668 · Azure/azure-sdk-pr (github.com)](https://github.com/Azure/azure-sdk-pr/issues/1668)

yifan-zhou922 committed:
Update models.tsp

yifan-zhou922 committed:
Update models.tsp

yifan-zhou922 committed:
Update readme.md

yifan-zhou922 committed:
Update readme.md

yifan-zhou922 committed:
update

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update openapi.json

yifan-zhou922 commented:
For the rest of the linkDiff warning, there are Atlas API format which we need to keep compatible.

chunyu3 commented:
hello @yifan-zhou922 The build failure in azure-sdk-for-net ci because it break csharp sdk Analyzer. You can ask archboard review and approve this, then suppress the ruler. otherwise, you need to customize for csharp.

yifan-zhou922 commented:
**Need archboard Review & Approve for the azure-sdk-for-net ci failure.**

1# We need the approval to keep the model names which end with "Options". These naming suggestions come from the SDK review meeting, but cause the failures to auto generate.net APIView.

AutoCompleteOptions.cs(13,26): error AZC0030: Model name 'AutoCompleteOptions' ends with 'Options'. Suggest to rename it to 'AutoCompleteConfig' or any other appropriate name.
ClassificationAssociateOptions.cs(14,26): error AZC0030: Model name 'ClassificationAssociateOptions' ends with 'Options'. Suggest to rename it to 'ClassificationAssociateConfig' or any other appropriate name.
MoveEntitiesOptions.cs(14,26): error AZC0030: Model name 'MoveEntitiesOptions' ends with 'Options'. Suggest to rename it to 'MoveEntitiesConfig' or any other appropriate name.
/QueryOptions.cs(15,26): error AZC0030: Model name 'QueryOptions' ends with 'Options'. Suggest to rename it to 'QueryConfig' or any other appropriate name. 
SuggestOptions.cs(13,26): error AZC0030: Model name 'SuggestOptions' ends with 'Options'. Suggest to rename it to 'SuggestConfig' or any other appropriate name. 
2# We name several operation groups like Discovery/Entity/Relationship/Lineage/etc. This also causes the build failures. However, these issues do not exist in other languages

Generated/Discovery.cs(19,26): error AZC0012: Single word class names are too generic and have high chance of collision with BCL types or types from other libraries 
Generated/Entity.cs(21,26): error AZC0012: Single word class names are too generic and have high chance of collision with BCL types or types from other libraries
/Glossary.cs(21,26): error AZC0012: Single word class names are too generic and have high chance of collision with BCL types or types from other libraries 
Lineage.cs(19,26): error AZC0012: Single word class names are too generic and have high chance of collision with BCL types or types from other libraries
 
Relationship.cs(19,26): error AZC0012: Single word class names are too generic and have high chance of collision with BCL types or types from other libraries 
 
Please help approve/suppress the analyzer rule or suggest the way to avoid the issue.

yifan-zhou922 commented:
> hello @yifan-zhou922 The build failure in azure-sdk-for-net ci because it break csharp sdk Analyzer. You can ask archboard review and approve this, then suppress the ruler. otherwise, you need to customize for csharp.

Thanks @chunyu3 , do you know who should I @ for the approval? Previously, according to @KrzysztofCwalina 's suggestion, the static anaylzer should not be complaining about the type names. And, I have also filed a bug https://github.com/Azure/azure-sdk-tools/issues/7589 as suggested.

mikeharder commented on custom-words.txt:
```diff
@@ -2912,6 +2912,7 @@ timeseries
 timeseriesinsights
 timespan
 timestep
+timerange
```
Please add this to `cSpell.json` instead of `custom-words.txt`

mikeharder commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/openapi.json:
Why are you adding a second OpenAPI file, rather than updating the existing file?

yifan-zhou922 replied:
The plan is to replace the existing purviewdatamap.json with the typeSpec one. 

As we have released the GA version and are also releasing another public preview API based on the current purviewdatamap.json [Entity - List By Guids - REST API (Azure Purview) | Microsoft Learn](https://learn.microsoft.com/en-us/rest/api/purview/datamapdataplane/entity/list-by-guids?view=rest-purview-datamapdataplane-2023-10-01-preview&tabs=HTTP), we choose to add a new file for this typeSpec migration instead of directly updating the existing file to avoid blocking the current documentation behavior. 

If it is not a blocker to merge this PR, we can continue releasing the SDKs based on typeSpec and later on replace the purviewdatamap.json.

mikeharder replied:
@allenjzhang: Any thoughts on this migration plan?

yifan-zhou922 replied:
We plan to update the existing file under 2023-09-01 to avoid the api doc generation pipeline failures. Thanks @catalinaperalta for help testing.

yifan-zhou922 committed:
move to cSpell.json

yifan-zhou922 committed:
Merge branch 'main' into feature/datamap-tsp

yifan-zhou922 committed:
update examples

yifan-zhou922 commented:
/azp run

yifan-zhou922 commented:
/azp run

yifan-zhou922 committed:
update examples

yifan-zhou922 committed:
fix api doc preview

yifan-zhou922 committed:
update csharp naming

yifan-zhou922 commented:
/azp run

yifan-zhou922 commented:
/azp run

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 commented:
> ****UPDATES:**
> For the above issues which causing the .net generation pipeline failures, we rename the modes to xxxConfig as recommended. And we rename our operation group names to EntityOperations/DiscoveryOperations/etc to avoid the AZC0012 error.**

**UPDATES:
For the above issues which causing the .net generation pipeline failures, we rename the modes to xxxConfig as recommended. And we rename our operation group names to EntityOperations/DiscoveryOperations/etc to avoid the AZC0012 error.**

yifan-zhou922 committed:
keep purviewdatamap.json

yifan-zhou922 committed:
Update readme.md

yifan-zhou922 committed:
test for csharp operation group name

yifan-zhou922 committed:
Update routes.tsp

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+      TraitLocation.Response,
+      TraitContext.Action
+    >,
+  Traits & NoConditionalRequests & NoRepeatableRequests & NoClientRequestId,
```
The traits in this template look redundant @timotheeguerin is this definition correct?

catalinaperalta replied:
@yifan-zhou922 if you update line 51 to the following do your requests still look the same:
```suggestion
  Traits,
```

yifan-zhou922 replied:
Updated Thanks

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+  @doc("List entities in bulk identified by its GUIDs.")
+  @route("/atlas/v2/entity/bulk")
+  @get
+  listByGuids is Azure.Core.Foundations.Operation<
```
Does this operation return a paged response?

yifan-zhou922 replied:
No. This is also an Atlas API.

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+      #suppress "@azure-tools/typespec-azure-core/bad-record-type" "Should use unknown to model Object"
+      @doc("BusinessMetadata payload")
+      @body
+      body: Record<Record<unknown>>;
```
Based on the original swagger, this type shouldnt be nested. I saw that the BusinessMetadata payload was previously defined like this:
``` 
"BusinessMetadata": {
      "type": "object",
      "additionalProperties": {
        "type": "object"
      },
      "description": "BusinessMetadata"
    },
```

So this type should just be `Record<unknown>`, there are other instances of this nested Record type that should be fixed if that's the case

yifan-zhou922 replied:
For buisnessMetadata, the format is <String, Object>. This is also an Atlas concept. We try to model it as a strong type but fail to achieve that.

timotheeguerin replied:
I think `Record<Record<unknown>` is correct in this case. It was previously scoped to `type: object` and `additionalProperties` is true by default (now we have not really been respecting that in autorest)

timotheeguerin replied:
i think it depends on what will cause the breaking change in the SDK. The previous swagger might have been interpreted as `REcord<unknown>`

catalinaperalta replied:
It's a new SDK so if this is an improvement in the modeling then it's better to go with the updated type if that's what the service returns.

yifan-zhou922 replied:
This is a bug in the original swagger. The businessMetadata should be defined as Map<String, Map<String, Object>> instead of Map<String,Object>

catalinaperalta replied:
Although on line 736 the business metadata payload is typed as `Record<unknown>`, it looks like it's for the same type of response, is that correct?

yifan-zhou922 replied:
Yes. It is correct. the 736 is business metadata **attributes**, which is Map<String,Object>

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+      #suppress "@azure-tools/typespec-azure-core/bad-record-type" "Should use unknown to model Object"
+      @doc("Business metadata payload")
+      @body
+      body: Record<Record<unknown>>;
```
Another example of the nested record types

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+    AtlasErrorResponse
+  >;
+
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is Atlas API"
```
Nit this operation shouldnt need the suppression.

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/client.tsp:
```diff
@@ -0,0 +1,139 @@
+@operationGroup
+interface Glossary {
+  batchGet is PurviewDataMap.Glossary.list;
+  create is PurviewDataMap.Glossary.create;
```
Thanks. The architects gave the feedback for renaming list to batchGet, not mentioning the name being too generic. We can address them for the GA SDK. Maybe, it is not a blocker for the preview release.

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2245 @@
+import "@typespec/versioning";
+import "@typespec/rest";
+import "./models.tsp";
+import "./main.tsp";
```
Updated. Thanks.

yifan-zhou922 committed:
fix comments

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2243 @@
+example. qualifiedName can be changed to other unique attributes)
+""")
+      @query
+      `attr_N:qualifiedName`?: string;
```
Just a heads up this will probably not work as expected in the SDKs, I understand there are limitations in describing this API, how was this handled in the previous SDKs?

yifan-zhou922 replied:
In previous SDK, only qualifiedName is supported. Users cannot add multiple unique attributes.

catalinaperalta replied:
The attr_N prefix if valid for the service?

yifan-zhou922 replied:
Yes. It is valid. It is also an Atlas API. We need to keep compatible

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,2146 @@
+  ...AtlasStruct;
+
+  @doc("The GUID of the entity.")
+  entityGuid?: string;
```
ping on this

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,1972 @@
+}
+
+@doc("MoveEntitiesOptions")
+@projectedName("csharp", "MoveEntitiesConfig")
```
Make sure to check for all instances of SDK specific renames and move them to client.tsp. Here are some docs on how to do that: https://azure.github.io/typespec-azure/docs/howtos/DataPlane%20Generation%20-%20DPG/04renaming#model-names

yifan-zhou922 replied:
The MoveEntitiesOption is renamed according to the SDK review. However, this naming did not pass the .net analyzer. So , we only change the naming for csharp.

yifan-zhou922 replied:
Updated

catalinaperalta commented on specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/purviewdatamap.json:
```diff
@@ -214,1406 +283,1499 @@
+            "$ref": "#/parameters/Azure.Core.Foundations.ApiVersionParameter"
+          },
+          {
+            "name": "file",
```
This name changed from the previous swagger?

yifan-zhou922 replied:
We notice in python SDK that the name used here will be the default key name. As in the API, the key name is file, so we think it is better to rename to file.
![image](https://github.com/Azure/azure-rest-api-specs/assets/78060542/42c487ba-6f34-4eef-8ff0-ab62d1604cc1)
![image](https://github.com/Azure/azure-rest-api-specs/assets/78060542/19d59d64-e88c-4be5-b0aa-3ff2c2640395)

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 committed:
move to client.tsp

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
Update client.tsp

yifan-zhou922 committed:
update api version query parameter

yifan-zhou922 committed:
update api version query parameter

yifan-zhou922 committed:
update

yifan-zhou922 commented:
In our previous GA swagger, several Atlas APIs do not require api-version as the required parameter.

Updated in the routes.tsp to align with the previous GA document. Only selected APIs api-version as required parameters. Others will treat api-version as the optional paramter.

yifan-zhou922 committed:
set the body for label API to align with the GA swagger doc

yifan-zhou922 committed:
Update api version query parameter

yifan-zhou922 commented:
/azp run

yifan-zhou922 commented:
/azp run

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2355 @@
+import "@azure-tools/typespec-azure-core";
+import "@azure-tools/typespec-client-generator-core";
```
We shouldn't import "@azure-tools/typespec-client-generator-core" outside the client.tsp file.

yifan-zhou922 replied:
Removed.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2355 @@
+      collectionId?: string;
+
+      @doc("Atlas entity with extended information.")
+      @body
```
Why do we need @body?

yifan-zhou922 replied:
Thanks. Updated to suggested way.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,1963 @@
+@doc("Captures time-boundary details")
+model TimeBoundary {
+  @doc("The end of the time boundary.")
+  endTime?: string;
```
Are these strings or some well-known date/date-time/time formats (e.g. ISO8601)

yifan-zhou922 replied:
They are following the format as "yyyy/MM/dd HH:mm:ss"

yifan-zhou922 replied:
It seems like we do not support ISO8601 for datetime, right?
![image](https://github.com/Azure/azure-rest-api-specs/assets/78060542/61e2a3e7-df96-4ead-a292-fd98522a9938)

johanste replied:
ISO8601 is the default serialization for date time in json (more specifically, we recommend the [internet date/time format profile](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6)). That doesn't seem to be the format you describe above, though?

yifan-zhou922 replied:
Yes. Atlas defines the format as this. And we need to be compatible with Atlas.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2355 @@
+The qualified name of the entity. (This is only an example. qualifiedName can
+be changed to other unique attributes)
+""")
+      @query("attr:qualifiedName")
```
This does not look correct - isn't this supposed to be arbitrary key/value pairs? 
```suggestion
      @query("attr:qualifiedName")
```

yifan-zhou922 replied:
The is the current documented API behavior. https://learn.microsoft.com/en-us/rest/api/purview/datamapdataplane/entity/get-by-unique-attributes?view=rest-purview-datamapdataplane-2023-09-01&tabs=HTTP

catalinaperalta replied:
We have an issue open for this @johanste, I'm looking for it to link. This is currently at the same level of support as the team had with swagger.

catalinaperalta replied:
Couldnt find the issue Johan, going to ping some other tsp folks if they recall what happened with the original issue, otherwise I will open a new one.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2355 @@
+  TErrorResponse
+>;
+
+model OkResponse<T> {
```
Why do we need this model? Can we at least make it an alias?

yifan-zhou922 replied:
Make it to alias. Thanks.
We need to define the status code of an operation.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/main.tsp:
```diff
@@ -0,0 +1,44 @@
+}
+
+@doc("The Azure Active Directory OAuth2 Flow")
+model AuthToken
```
Why is this a model? I only see it  used as a parameter to the `@useAuth` decorator. So we should either define it inline (pass it to the decorator) or make it an alias.

yifan-zhou922 replied:
Updated.

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,1963 @@
+}
+
+@doc("The base model object.")
+model AtlasBaseModelObject {
```
Why does this object exist? I don't see it being used other than spread into AtlasGlossaryBaseObject. It most likely should either be an alias or not exist.

yifan-zhou922 replied:
Rename to alias. Thanks

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,1963 @@
+}
+
+@doc("Base class that captures common-attributes for all types.")
+model AtlasBaseTypeDef {
```
This should be an alias.

yifan-zhou922 replied:
Updated

johanste commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2355 @@
+  TParams extends TypeSpec.Reflection.Model,
+  TResponse,
+  Traits extends TypeSpec.Reflection.Model = {},
+  TErrorResponse = Azure.Core.Foundations.ErrorResponse
```
We should either default to `AtlasErrorResponse` or simply not expose this template parameter if there are no endpoints that need to override the default. I believe all endpoints are explicitly overriding it now, no?

yifan-zhou922 replied:
Thanks. All responses should use the AtlasErrorResponse. Define it into the default behavior.

yifan-zhou922 committed:
update for comments

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/client.tsp:
```diff
@@ -0,0 +1,161 @@
+})
+namespace Customizations;
+
+@@access(MoveEntitiesOptions, Access.public);
```
Thanks. Removed those decorators.

yifan-zhou922 committed:
update

yifan-zhou922 committed:
Update routes.tsp

yifan-zhou922 commented:
/azp run

yifan-zhou922 commented:
Because of the issue for csharp generator which in mentioned here https://github.com/Azure/autorest.csharp/issues/4189,
we will use int64 instead of utcDateTime in this TypeSpec migration. 
![image](https://github.com/Azure/azure-rest-api-specs/assets/78060542/3e62bcc5-e8a0-45a2-bffc-27b9d9f6d115)

yifan-zhou922 committed:
update the createTime/updateTIme for an issue in csharp generator

msyyc commented:
/azp run

mikeharder commented:
> Because of the issue for csharp generator which in mentioned here [Azure/autorest.csharp#4189](https://github.com/Azure/autorest.csharp/issues/4189), we will use int64 instead of utcDateTime in this TypeSpec migration. ![image](https://private-user-images.githubusercontent.com/78060542/302090535-3e62bcc5-e8a0-45a2-bffc-27b9d9f6d115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNzA2MTcsIm5iZiI6MTcwNzE3MDMxNywicGF0aCI6Ii83ODA2MDU0Mi8zMDIwOTA1MzUtM2U2MmJjYzUtZThhMC00NWEyLWJmZmMtMjdiOWQ5ZjZkMTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA1VDIxNTgzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1YzZmMTY0ODViMzkzOGQ4ODE0MDZkOWQxYTA5ZjMyMmQyMDgzOGQ4ZDg1Y2E1NjU1N2I2MGRlNzU2ZGEwMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rQ_Sm-YJJf2plesa27rx91sm_TdpuVnQBhF4mJTOCJA)

@yifan-zhou922: If you can wait, it would be nice to see if autorest.csharp can hotfix this quickly, so you don't need to workaround this in your spec.

mikeharder commented:
> Because of the issue for csharp generator which in mentioned here [Azure/autorest.csharp#4189](https://github.com/Azure/autorest.csharp/issues/4189), we will use int64 instead of utcDateTime in this TypeSpec migration. ![image](https://private-user-images.githubusercontent.com/78060542/302090535-3e62bcc5-e8a0-45a2-bffc-27b9d9f6d115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNzg1NzUsIm5iZiI6MTcwNzE3ODI3NSwicGF0aCI6Ii83ODA2MDU0Mi8zMDIwOTA1MzUtM2U2MmJjYzUtZThhMC00NWEyLWJmZmMtMjdiOWQ5ZjZkMTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDAwMTExNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3ZWI1NDFmY2Y0ZTFjZDQwYjNjZTA4NmE5ZTUyYTNmNzIzYjlhY2VkNmE1MzkyMjk5ZjUyMGZhMzdlYTA4ZWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MjUiOOuX1_H2w80kRLV7RXrCaYRGbFE-0fdyKr44DuU)

Encoding `unixTimestamp` is intended to represent **seconds** since epoch, not **milliseconds**.  Please see discussion here:

https://github.com/microsoft/typespec/issues/2887

Longer-term, it sounds like we might add a dedicated type for this to TypeSpec.  Until then, we recommend modeling this as a `safeint` instead of an `int64`, since `safeint` is more compatible and large enough to hold very large dates.

yifan-zhou922 committed:
Revert "update the createTime/updateTIme for an issue in csharp generator"

This reverts commit a9e8ec26f7d01788c8b94c582a7de0b82c00ce66.

yifan-zhou922 committed:
update with safeint

yifan-zhou922 commented:
Thanks for the comments. Updated to safeint.

catalinaperalta commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2133 @@
+      guid: string;
+
+      @doc("An array of classifications to be added.")
+      @body
+      body: AtlasClassification[];
```
Looks like some of the `@body` instances are still pending an update to having the operations parameters intersect with the request body. Should be the same as the changes you added above.

yifan-zhou922 replied:
![image](https://github.com/Azure/azure-rest-api-specs/assets/78060542/0c8ba781-36fc-41b9-bd8f-da1d7f6ba372)
I have tried before but having an error here.

catalinaperalta replied:
Ah I see, thanks for sharing the info

yifan-zhou922 committed:
Merge branch 'main' into feature/datamap-tsp

yifan-zhou922 committed:
update to encodeName

yifan-zhou922 commented on specification/purview/Azure.Analytics.Purview.DataMap/models.tsp:
```diff
@@ -0,0 +1,1949 @@
+the field, attached with emphasis mark, is returned as an element of
+@search.highlights.
+""")
+  @projectedName("json", "@search.highlights")
```
Updated

catalinaperalta commented:
Posting the comments related to the CI failures here since they've gotten lost in the long thread of this PR:
- Python failures are due to an emitter issue that @msyyc is tracking here: https://github.com/Azure/typespec-azure/issues/215 
- .NET issues are due to some naming issues that need to be resolved in the .NET repo when merging in the SDK, related comments with a link to some historical comments about this issue from @yifan-zhou922 (cc @chunyu3 ):
> For operation group name (AZC0012), we want to add a suppression rule like this when releasing the SDK. Prefer not to block this PR just for this failure. [Leo Lie: azure-sdk-for-net-track2: How to suppress AZC0012: Single word class names?](https://teams.microsoft.com/l/message/19:9efda948665d481086766c0b7572abc3@thread.skype/1699993167104?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1699993167104&teamName=Azure%20SDK&channelName=Service%20API%20Toolset&createdTime=1699993167104)
posted in Azure SDK / Service API Toolset on 2023年11月15日 04:19

yifan-zhou922 commented:
/pr RequestMerge

SerenaLin2020 commented on specification/purview/Azure.Analytics.Purview.DataMap/routes.tsp:
```diff
@@ -0,0 +1,2133 @@
+    {
+      @doc("BusinessMetadata name")
+      @path
+      businessMetadataName: string;
```
GA version upgrade to TypeSpec
